# Hola Juan!

Mi nombre es David Bautista, soy code reviewer de Tripleten y hoy tengo el gusto de revisar tu proyecto.

Cuando vea un error laa primera vez, lo señalaré. Deberás encontrarlo y arreglarlo. La intención es que te prepares para un espacio real de trabajo. En un trabajo, el líder de tu equipo hará lo mismo. Si no puedes solucionar el error, te daré más información en la próxima ocasión.

Encontrarás mis comentarios más abajo - por favor, no los muevas, no los modifiques ni los borres.

¿Cómo lo voy a hacer? Voy a leer detenidamente cada una de las implementaciones que has llevado a cabo para cumplir con lo solicitado. Verás los comentarios de esta forma:

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Si todo está perfecto.
</div>


<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Si tu código está bien pero se puede mejorar o hay algún detalle que le hace falta.
</div>


<div class="alert alert-block alert-danger">
    
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Si de pronto hace falta algo o existe algún problema con tu código o conclusiones.
</div>


Puedes responderme de esta forma: 

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>
</div>

¡Empecemos!

# Déjame escuchar la música

# Contenido <a id='back'></a>

* [Introducción](#intro)
* [Etapa 1. Descripción de los datos](#data_review)
    * [Conclusiones](#data_review_conclusions)
* [Etapa 2. Preprocesamiento de datos](#data_preprocessing)
    * [2.1 Estilo del encabezado](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusiones](#data_preprocessing_conclusions)
* [Etapa 3. Prueba de hipótesis](#hypothesis)
    * [3.1 Hipótesis 1: actividad de los usuarios y las usuarias en las dos ciudades](#activity)
* [Conclusiones](#end)

## Introducción <a id='intro'></a>
Como analista de datos, tu trabajo consiste en analizar datos para extraer información valiosa y tomar decisiones basadas en ellos. Esto implica diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Siempre que investigamos, necesitamos formular hipótesis que después podamos probar. A veces aceptamos estas hipótesis; otras veces, las rechazamos. Para tomar las decisiones correctas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

En este proyecto, compararás las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiarás datos reales de transmisión de música online para probar la hipótesis a continuación y comparar el comportamiento de los usuarios y las usuarias de estas dos ciudades.

### Objetivo:
Prueba la hipótesis:
1. La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.


### Etapas
Los datos del comportamiento del usuario se almacenan en el archivo `/datasets/music_project_en.csv`. No hay ninguna información sobre la calidad de los datos, así que necesitarás examinarlos antes de probar la hipótesis.

Primero, evaluarás la calidad de los datos y verás si los problemas son significativos. Entonces, durante el preprocesamiento de datos, tomarás en cuenta los problemas más críticos.

Tu proyecto consistirá en tres etapas:
 1. Descripción de los datos.
 2. Preprocesamiento de datos.
 3. Prueba de hipótesis.








[Volver a Contenidos](#back)

## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
**Comentario Importante**
    
Hola, Juan, recuerda que el ideal es no eliminar los comentarios de revisión para dar una mejor continuidad.  </div>

Necesitarás `pandas`, así que impórtalo.

In [2]:
import pandas as pd #utilizo import para importar la libreria de pandas

<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
Perfecto, muy buen trabajo importando ``pandas``.</div>

Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [3]:
df = pd.read_csv('/datasets/music_project_en.csv') #utilizo la funcion read para leer el archivo csv

<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
Buen trabajo cargando los datos necesarios para el desarrollo del proyecto.⁣</div>

Muestra las 10 primeras filas de la tabla:

In [4]:
df.head(10) #imprime las 10 primeras filas


userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5  Shelbyville  13:09:41     Friday  
6  Springfield  13:00:07  Wednesday  
7  Springfield  20:47:49  Wednesday  
8  Springfield  09:17:40     Friday  
9  Shelbyville  21:20:49  Wednesday

<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
Buen trabajo con el uso del método ``head()``.</div>

Obtén la información general sobre la tabla con un comando. Conoces el método que muestra la información general que necesitamos.

In [5]:
df.info() # Obtener la información general sobre nuestros datos


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
Buen trabajo con el uso del método ``info()``.</div>

Estas son nuestras observaciones sobre la tabla. Contiene siete columnas. Almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario o la usuaria;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la pista;
- `'City'`: ciudad del usuario o la usuaria;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. Hay espacios en algunos encabezados.
3. El tercer problema es en ' userID' ya que le falta un caracter '_' por ejemplo: 'user_id'.




### Escribe observaciones de tu parte. Estas son algunas de las preguntas que pueden ser útiles: <a id='data_review_conclusions'></a>

`1.   ¿Qué tipo de datos tenemos a nuestra disposición en las filas? ¿Y cómo podemos entender lo que almacenan las columnas?`

Tenemos strings, intregers, floats. Podemos entender los datos haciendo un análisis de lo que los datos nos están contando tambien revisando de que se trata en los encabezados.

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`

Yo creo que si tenemos suficientes datos para llegar a una conclusión más sin embargo siempre es ideal ver si se pueden obtener más datos ya que mientras más datos hayan más preciso es la conclusión.

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`
Sí, noté varios problemas como datos duplicados y ausentes. Tambien logré ver que los encabezados no cumplen con el snake_case.

<div class="alert alert-block alert-danger">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
~~No olvides generar observaciones basándote en las incógnitas planteadas.~~⁣ </div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor #3</b> <a class="tocSkip"></a>
    
Buen trabajo Juan. </div>

[Volver a Contenidos](#back)

## Etapa 2. Preprocesamiento de datos <a id='data_preprocessing'></a>

El objetivo aquí es preparar los datos para que sean analizados.
El primer paso es resolver cualquier problema con los encabezados. Luego podemos avanzar a los valores ausentes y duplicados. Empecemos.

Corrige el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Muestra los encabezados de la tabla (los nombres de las columnas):

In [6]:
column_names = df.columns # este código muestra los nombres de las columnas.
print(column_names)

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
Bien hecho, buen trabajo mostrando cómo se ve cada uno de los encabezados de la tabla.</div>

Cambia los encabezados de la tabla de acuerdo con las reglas del buen estilo:
* Todos los caracteres deben ser minúsculas.
* Elimina los espacios.
* Si el nombre tiene varias palabras, utiliza snake_case.

Anteriormente, aprendiste acerca de la forma automática de cambiar el nombre de las columnas. Vamos a aplicarla ahora. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [7]:
column_names = df.columns
column_names = [column.lower() for column in df.columns] # Bucle en los encabezados poniendo todo en minúsculas
print(column_names)

['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day']


Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas e imprime los nombres de las columnas nuevamente:

In [8]:
column_names = df.columns      # Bucle en los encabezados eliminando los espacios
for columns in df.columns:
    names_lower = columns.lower()
    names_strip = names_lower.strip()
    df.rename(columns={columns: names_strip}, inplace=True)
print(column_names)

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
Perfecto, buen trabajo con el bucle usado para aplicar tanto el lowering como el strip a cada encabezado.</div>

Necesitamos aplicar la regla de snake_case a la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [9]:
column_names = df.columns
df.rename(columns={'userid': 'user_id'}, inplace=True) # Cambiar el nombre de la columna "userid"
print(column_names)

Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
Perfecto, buen trabajo aplicando el formaro *snake_case* al encabezado ``userid``.</div>

Comprueba el resultado. Muestra los encabezados una vez más:

In [10]:
print(column_names)# Comprobar el resultado: la lista de encabezados


Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


[Volver a Contenidos](#back)

### Valores ausentes <a id='missing_values'></a>
 Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos en una secuencia para obtener el número de valores ausentes.

In [11]:
num_values = df.isna().sum() # Calcular el número de valores ausentes
print(num_values) 


user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
Buen trabajo mostrando la cantidad de valores nulos de cada una de las columnas.⁣</div>

No todos los valores ausentes afectan a la investigación. Por ejemplo, los valores ausentes en `track` y `artist` no son cruciales. Simplemente puedes reemplazarlos con valores predeterminados como el string `'unknown'` (desconocido).

Pero los valores ausentes en `'genre'` pueden afectar la comparación entre las preferencias musicales de Springfield y Shelbyville. En la vida real, sería útil saber las razones por las cuales hay datos ausentes e intentar recuperarlos. Pero no tenemos esa oportunidad en este proyecto. Así que tendrás que:
* rellenar estos valores ausentes con un valor predeterminado;
* evaluar cuánto podrían afectar los valores ausentes a tus cómputos;

Reemplazar los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`. Como mostramos anteriormente en las lecciones, la mejor forma de hacerlo es crear una lista que almacene los nombres de las columnas donde se necesita el reemplazo. Luego, utiliza esta lista e itera sobre las columnas donde se necesita el reemplazo haciendo el propio reemplazo.

In [12]:
columns_fill = ['track', 'artist', 'genre']

for column in columns_fill:  # Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
    df[column].fillna('unknown', inplace=True)
print(df) 


        user_id                              track            artist  \
0      FFB692EC                  Kamigata To Boots  The Mass Missile   
1      55204538        Delayed Because of Accident  Andreas Rönnberg   
2        20EC38                  Funiculì funiculà       Mario Lanza   
3      A3DD03C9              Dragons in the Sunset        Fire + Ice   
4      E2DC1FAE                        Soul People        Space Echo   
...         ...                                ...               ...   
65074  729CBB09                            My Name            McLean   
65075  D08D4A55  Maybe One Day (feat. Black Spade)       Blu & Exile   
65076  C5E3A0D5                          Jalopiina           unknown   
65077  321D0506                      Freight Train     Chas McDevitt   
65078  3A64EF84          Tell Me Sweet Little Lies      Monica Lopez   

            genre         city      time        day  
0            rock  Shelbyville  20:28:33  Wednesday  
1            rock  Springfi

<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
Buen trabajo solucionando el problema de los nulos en la tabla. Como consejo, podrías tener en cuenta no imprimir el dataframe en cada iteración del bucle para evitar ruido innecesario.⁣⁣⁣ </div>

Ahora comprueba el resultado para asegurarte de que después del reemplazo no haya valores ausentes en el conjunto de datos. Para hacer esto, cuenta los valores ausentes nuevamente.

In [13]:
print(df.isna().sum()) # Contar valores ausentes


user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


<div class="alert alert-block alert-danger">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
~~Fíjate que al mostrar la tabla pareciera que no se hubiera solucionado el problema de los nulos, pero eso sucede porque la variable con el conteo lo habías realizado antes de eliminar los nulos. Puedes aplicar todo el método otra vez para que ahora veas la cifra actualizada. ~~ </div>

<div class="alert alert-block alert-info">
<b>Ejecuté el código de nuevo sin guardar en variable pero arrojó lo mismo.</b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor #3</b> <a class="tocSkip"></a>
    
Volvi a ejecutar el proyecto y funciono Juan. Buen trabajo. </div>

[Volver a Contenidos](#back)

### Duplicados <a id='duplicates'></a>
Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos en una secuencia para obtener la cantidad de duplicados explícitos.

In [14]:
values_duplicated = df.duplicated().sum()
print(values_duplicated) # Contar duplicados explícitos


3826


<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
Buen trabajo identificando la cantidad de registros duplicados dentro de la tabla.</div>

Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [16]:
df= df.drop_duplicates().reset_index(drop=True) # Eliminar duplicados explícitos


Comprobemos ahora si eliminamos con éxito todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [17]:
print(df.duplicated().sum()) # Comprobar de nuevo si hay duplicados



0


<div class="alert alert-block alert-danger">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    

~~Recuerda que una vez se eliminan duplicados explícitos, lo más aconsejable es que debes reindexar la tabla para no tener problemas con esto en un futuro, lo puedes hacer así:~~
```python
df = df.drop_duplicates().reset_index(drop=True)
```
    
~~Por otro lado, es muy importante que no crees otra tabla, sino que la eliminación de los duplicados que guardad en ``df``, para que el resto del proyectos se desarrolle con los datos limpios.~~
</div>

<div class="alert alert-block alert-info">
<b>Espero haber hecho la correcición con éxito.</b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor #3</b> <a class="tocSkip"></a>
    
~~Recuerda que la idea es guardar los cambios en el mismo dataframe, para esto puedes asignar el resultado del método a la misma variable, sigue tal cual la estructura que te había mostrado en la revisión anterior.~~  </div>

<div class="alert alert-block alert-info">
<b>Listo, corregido!.</b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor #4</b> <a class="tocSkip"></a>
    
Perfecto Juan, muy buen trabajo.</div>

Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Para hacerlo, primero mostremos una lista de nombres de género únicos, ordenados en orden alfabético. Para ello:
* Extrae la columna `genre` del DataFrame.
* Llama al método que devolverá todos los valores únicos en la columna extraída.


In [29]:
unique_values = df['genre'].dropna().unique() # Inspeccionar los nombres de géneros únicos

unique_values.sort()
print(unique_values) 


['acid' 'acoustic' 'action' 'adult' 'africa' 'afrikaans' 'alternative'
 'ambient' 'americana' 'animated' 'anime' 'arabesk' 'arabic' 'arena'
 'argentinetango' 'art' 'audiobook' 'avantgarde' 'axé' 'baile' 'balkan'
 'beats' 'bigroom' 'black' 'bluegrass' 'blues' 'bollywood' 'bossa'
 'brazilian' 'breakbeat' 'breaks' 'broadway' 'cantautori' 'cantopop'
 'canzone' 'caribbean' 'caucasian' 'celtic' 'chamber' 'children' 'chill'
 'chinese' 'choral' 'christian' 'christmas' 'classical' 'classicmetal'
 'club' 'colombian' 'comedy' 'conjazz' 'contemporary' 'country' 'cuban'
 'dance' 'dancehall' 'dancepop' 'dark' 'death' 'deep' 'deutschrock'
 'deutschspr' 'dirty' 'disco' 'dnb' 'documentary' 'downbeat' 'downtempo'
 'drum' 'dub' 'dubstep' 'eastern' 'easy' 'electronic' 'electropop' 'emo'
 'entehno' 'epicmetal' 'estrada' 'ethnic' 'eurofolk' 'european'
 'experimental' 'extrememetal' 'fado' 'film' 'fitness' 'flamenco' 'folk'
 'folklore' 'folkmetal' 'folkrock' 'folktronica' 'forró' 'frankreich'
 'französisch' 

<div class="alert alert-block alert-danger">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
~~Recuerda que el ideal en la sección es mostrar los valores únicos de la columna, pero de manera ordenada alfabéticamente.~~ </div>

<div class="alert alert-block alert-info">
<b>Espero haber hecho la corrección con éxito.</b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor #3</b> <a class="tocSkip"></a>
    
Perfecto, buen trabajo Juan. </div>

Busca en la lista para encontrar duplicados implícitos del género `hiphop`. Estos pueden ser nombres escritos incorrectamente o nombres alternativos para el mismo género.

Verás los siguientes duplicados implícitos:
* `hip`
* `hop`
* `hip-hop`

Para deshacerte de ellos, crea una función llamada `replace_wrong_genres()` con dos parámetros:
* `wrong_genres=`: esta es una lista que contiene todos los valores que necesitas reemplazar.
* `correct_genre=`: este es un string que vas a utilizar como reemplazo.

Como resultado, la función debería corregir los nombres en la columna `'genre'` de la tabla `df`, es decir, remplazar cada valor de la lista `wrong_genres` por el valor en `correct_genre`.

Dentro del cuerpo de la función, utiliza un bucle `'for'` para iterar sobre la lista de géneros incorrectos, extrae la columna `'genre'` y aplica el método `replace` para hacer correcciones.

In [30]:
def replace_wrong_genres(df, wrong_genres, correct_genre): # Función para reemplazar duplicados implícitos

    for wrong_genre in wrong_genres:
        df['genre'] = df['genre'].replace(wrong_genres, correct_genre)

Ahora, llama a `replace_wrong_genres()` y pásale tales argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los reemplace por `hiphop`:

In [31]:
wrong_genres = ['hip','hop','hip-hop']  # Eliminar duplicados implícitos

correct_genre = 'hiphop'
                                          
replace_wrong_genres(df, wrong_genres, correct_genre)
print(df)


        user_id                              track            artist  \
0      FFB692EC                  Kamigata To Boots  The Mass Missile   
1      55204538        Delayed Because of Accident  Andreas Rönnberg   
2        20EC38                  Funiculì funiculà       Mario Lanza   
3      A3DD03C9              Dragons in the Sunset        Fire + Ice   
4      E2DC1FAE                        Soul People        Space Echo   
...         ...                                ...               ...   
65074  729CBB09                            My Name            McLean   
65075  D08D4A55  Maybe One Day (feat. Black Spade)       Blu & Exile   
65076  C5E3A0D5                          Jalopiina           unknown   
65077  321D0506                      Freight Train     Chas McDevitt   
65078  3A64EF84          Tell Me Sweet Little Lies      Monica Lopez   

            genre         city      time        day  
0            rock  Shelbyville  20:28:33  Wednesday  
1            rock  Springfi

<div class="alert alert-block alert-warning">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
~~Creaste bien la función para solucionar el problema de los duplicados implícitos dentro de la columna; sin embargo, trata de evitar duplicar el código de aplicación de la misma, por otro lado, no es tan necesario que imprimar tantas veces el dataframe, esto solo causa ruido en el contenido del proyecto.~~
</div>

<div class="alert alert-block alert-info">
<b>Espero haber hecho la corrección con éxito.</b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor #3</b> <a class="tocSkip"></a>
    
Perfecto, muy buen trabajo Juan.  </div>

Asegúrate de que los nombres duplicados han sido eliminados. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [32]:
unique_values = df['genre'].dropna().unique()  # Comprobación de duplicados implícitos
unique_values.sort()
print(unique_values) 


['acid' 'acoustic' 'action' 'adult' 'africa' 'afrikaans' 'alternative'
 'ambient' 'americana' 'animated' 'anime' 'arabesk' 'arabic' 'arena'
 'argentinetango' 'art' 'audiobook' 'avantgarde' 'axé' 'baile' 'balkan'
 'beats' 'bigroom' 'black' 'bluegrass' 'blues' 'bollywood' 'bossa'
 'brazilian' 'breakbeat' 'breaks' 'broadway' 'cantautori' 'cantopop'
 'canzone' 'caribbean' 'caucasian' 'celtic' 'chamber' 'children' 'chill'
 'chinese' 'choral' 'christian' 'christmas' 'classical' 'classicmetal'
 'club' 'colombian' 'comedy' 'conjazz' 'contemporary' 'country' 'cuban'
 'dance' 'dancehall' 'dancepop' 'dark' 'death' 'deep' 'deutschrock'
 'deutschspr' 'dirty' 'disco' 'dnb' 'documentary' 'downbeat' 'downtempo'
 'drum' 'dub' 'dubstep' 'eastern' 'easy' 'electronic' 'electropop' 'emo'
 'entehno' 'epicmetal' 'estrada' 'ethnic' 'eurofolk' 'european'
 'experimental' 'extrememetal' 'fado' 'film' 'fitness' 'flamenco' 'folk'
 'folklore' 'folkmetal' 'folkrock' 'folktronica' 'forró' 'frankreich'
 'französisch' 

<div class="alert alert-block alert-danger">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
~~Nuevamente, recuerda que el ideal en la sección es mostrar los valores únicos de la columna, pero de manera ordenada alfabéticamente.~~ </div>

<div class="alert alert-block alert-info">
<b>Espero haber hecho la corrección con éxito.</b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor #3</b> <a class="tocSkip"></a>
    
Perfecto!! </div>

[Volver a Contenidos](#back)

### Tus observaciones <a id='data_preprocessing_conclusions'></a>

`Al limpiar los datos y eliminar duplicados, hemos hecho que  nuestro análisis sea más confiable. Este proceso nos ayudó a asegurar que los datos eran precisos, lo cual confirmamos al revisar las tablas limpias. Gracias a esto, podemos identificar tendencias y patrones con mayor claridad, evitando errores causados por información redundante. En conclusión, esta limpieza fue un paso crucial para obtener resultados más precisos y comprensibles para nuestro análisis que se terminará de mostrar más abajo mostrando resultados con los datos ya limpios!`

<div class="alert alert-block alert-danger">
<b>Comentario del revisor #3</b> <a class="tocSkip"></a>
    
~~Ya que se han modificado algunas secciones del proyecto, sería interesante generar unas observaciones mejor estructuradas. ~~</div>

<div class="alert alert-block alert-info">
<b>Espero haber hecho la corrección con éxito!.</b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor #4</b> <a class="tocSkip"></a>
    
Muy buen trabajo, Juan, gracias por tener en cuenta todos mis comentarios. </div>

[Volver a Contenidos](#back)

## Etapa 3. Prueba de hipótesis <a id='hypothesis'></a>

### Hipótesis: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id='activity'></a>

La hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes.

* Agrupa a los usuarios y las usuarias por ciudad.
* Compara el número de canciones que cada grupo reprodujo el lunes, el miércoles y el viernes.


Realiza cada cálculo por separado.

El primer paso es evaluar la actividad del usuario en cada ciudad. Recuerda las etapas dividir-aplicar-combinar de las que hablamos anteriormente en la lección. Tu objetivo ahora es agrupar los datos por ciudad, aplicar el método apropiado para contar durante la etapa de aplicación y luego encontrar la cantidad de canciones reproducidas en cada grupo especificando la columna para obtener el recuento.

A continuación se muestra un ejemplo de cómo debería verse el resultado final:
`df.groupby(by='....')['column'].method()`Realiza cada cálculo por separado.

Para evaluar la actividad de los usuarios y las usuarias en cada ciudad, agrupa los datos por ciudad y encuentra la cantidad de canciones reproducidas en cada grupo.



In [33]:
track_count = df.groupby(by='city')['track'].count()# Contar las canciones reproducidas por ciudad
print(track_count)


city
Shelbyville    19719
Springfield    45360
Name: track, dtype: int64


`Existen muchos datos duplicados`

Ahora agrupemos los datos por día de la semana y encontremos el número de canciones reproducidas el lunes, miércoles y viernes. Utiliza el mismo método que antes, pero ahora necesitamos una agrupación diferente.


In [34]:
track_count = df.groupby(by='day')['track'].count()# Contar las canciones reproducidas por dia
print(track_count)

day
Friday       23149
Monday       22697
Wednesday    19233
Name: track, dtype: int64


<div class="alert alert-block alert-danger">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
~~Buen trabajo con la primera agrupación. Para el caso de la segunda, en la que se trata de ver la cantidad de canciones reproducidas por día, lo aconsejable sería seguir el primer ejemplo que hiciste por ciudades simplemente cambiando la variable.~~ </div>

<div class="alert alert-block alert-info">
<b>Espero haber hecho la corrección con éxito.</b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor #3</b> <a class="tocSkip"></a>
    
Perfecto, muy buen trabajo Juan.</div>

`El viernes se escucharon más canciones`

Ya sabes cómo contar entradas agrupándolas por ciudad o día. Ahora necesitas escribir una función que pueda contar entradas según ambos criterios simultáneamente.

Crea la función `number_tracks()` para calcular el número de canciones reproducidas en un determinado día **y** ciudad. La función debe aceptar dos parámetros:

- `day`: un día de la semana para filtrar. Por ejemplo, `'Monday'` (lunes).
- `city`: una ciudad para filtrar. Por ejemplo, `'Springfield'`.

Dentro de la función, aplicarás un filtrado consecutivo con indexación lógica.

Primero filtra los datos por día y luego filtra la tabla resultante por ciudad.

Después de filtrar los datos por dos criterios, cuenta el número de valores de la columna 'user_id' en la tabla resultante. Este recuento representa el número de entradas que estás buscando. Guarda el resultado en una nueva variable y devuélvelo desde la función.

In [35]:
def number_tracks(df, day, city): # Declara la función number_tracks() con dos parámetros: day= y city=.
    filtered_by_day = df[df['day'] == day] # Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=
    filtered_by_day_and_city = filtered_by_day[filtered_by_day['city'] == city] # Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=
    track_count = filtered_by_day_and_city['user_id'].count() # Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()
    return track_count # Devulve el número de valores de la columna 'user_id'
   

Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que recuperes los datos de ambas ciudades para cada uno de los tres días.

In [36]:
springfield_monday = number_tracks(df, 'Monday', 'Springfield') # El número de canciones reproducidas en Springfield el lunes
print(springfield_monday)

16715


In [37]:
shelbyville_monday = number_tracks(df, 'Monday', 'Shelbyville') # El número de canciones reproducidas en Shelbyville el lunes
print(shelbyville_monday)

5982


In [38]:
springfield_wednesday = number_tracks(df, 'Wednesday', 'Springfield') # El número de canciones reproducidas en Springfield el miércoles
print(springfield_wednesday)

11755


In [39]:
shelbyville_wednesday = number_tracks(df, 'Wednesday', 'Shelbyville') # El número de canciones reproducidas en Shelbyville el miércoles
print(shelbyville_wednesday)

7478


In [40]:
springfield_friday = number_tracks(df, 'Friday', 'Springfield') # El número de canciones reproducidas en Springfield el viernes
print(springfield_friday)

16890


In [41]:
shelbyville_friday = number_tracks(df, 'Friday', 'Shelbyville') # El número de canciones reproducidas en Shelbyville el viernes
print(shelbyville_friday)

6259


**Conclusiones**

`La hipotesis es correcta en springfield se escucha más musica que en Shelbyville, creo los datos fueron suficientes para este análisis y pienso que fue un caso, muy interesante!`

[Volver a Contenidos](#back)

# Conclusiones <a id='end'></a>

`El análisis y limpieza de los datos nos muestra una clara diferencia en que ciudad escucha más música entre Shelbyville y Springfield. En Shelbyville, se registraron 19,719 canciones, mientras que en Springfield la cifra es significativamente mayor, con un total de 45,360 canciones. Esta diferencia nos dice que los habitantes de Springfield escuchan más música que en Shelbyville. Esto puede significar más disponibilidad de tiempo o de acceso a medios musicales entre las dos ciudades.!`

<div class="alert alert-block alert-danger">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
~~Buen trabajo con la última sección; sin embargo, falta mejorar las conclusiones tanto de dicho apartado como las generales del proyecto.~~ </div>


<div class="alert alert-block alert-info">
<b>Espero haber hecho la corrección con éxito.</b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor #3</b> <a class="tocSkip"></a>
    
Siento que se podrían generar conclusiones un poco mejores estructuradas; sin embargo, muy buen trabajo. </div>

<div class="alert alert-block alert-info">
<b>Intentaré mejorar mis conclusiones, para este proyecto, siento que no tengo tantas cosas que decir ya que los datos son muy claros, pero tomaré tu recomendación en consideración, gracias!.</b> <a class="tocSkip"></a>
</div>

### Nota
En proyectos de investigación reales, la prueba de hipótesis estadística es más precisa y cuantitativa. También ten en cuenta que no siempre se pueden sacar conclusiones sobre una ciudad entera a partir de datos de una sola fuente.

Aprenderás más sobre la prueba de hipótesis en el sprint de análisis estadístico de datos.

<div class="alert alert-block alert-danger">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>

# Comentario General #2

~~Hola Juan, he dejado comentarios para que pueda tenerlos en cuenta para la siguiente entrega.~~  </div>

<div class="alert alert-block alert-info">
<b> Muchas gracias David!
Espero haber hecho las correciones con éxito, te pido una disculpa por el incoveniente anterior pero de verdad fue sin querer eliminé todo en así tuve que devolvertelo, espero haberlo hecho mejor y espero tu retroalimentación si algo tengo que cambiar, gracias!.</b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor #3</b> <a class="tocSkip"></a>

# Comentario General #3

~~Hola, Juan, te felicito por los cambios generados hasta el momento. Muy buen trabajo. Ahora bien, he dejado unos nuevos comentarios para que puedas tener en cuenta.~~  </div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor #3</b> <a class="tocSkip"></a>

# Comentario General #3

Hola, Juan, te felicito por la culminación del proyecto, muy buen trabajo.  </div>

[Volver a Contenidos](#back)